Feat (Core): ImportUtil refactor to improve responses

jgambarios opened the pull request:
This PR represents the final stage of the Import Content response improvements, where the `ImportContentletsProcessor` now leverages the new structured response format while maintaining full backward compatibility with existing Struts implementations.

### Changes Overview

* Updated `ImportContentletsProcessor` to use new response format 
* Added converter to maintain backward compatibility with Struts actions
* Improved error handling and validation message organization
* Enhanced documentation and code organization

### What's Not Changed

We intentionally kept several aspects unchanged to ensure stability:
* Core import logic remains untouched
* Legacy response format is still fully supported
* Existing workflow processing behavior
* Current transaction handling
* Permission checking mechanisms

While working on this code, we identified several potential improvements that were intentionally left for future iterations to keep the scope focused on the immediate needs.

This PR is related to: #30882
This PR fixes: #31339

This PR fixes: #31339

jgambarios committed:
#31339 Refactor import validation logic and improve JSON mappings

Updated import-related classes to enhance JSON serialization/deserialization, removed unused enums, added new validation codes, and streamlined legacy format conversion in `ImportResultConverter`. These changes improve code maintainability and align with new API requirements.

jgambarios committed:
Merge remote-tracking branch 'origin/main' into issue-31339-Migrate-ImportContentletsProcessor-to-Use-New-Import-Result-Format-While-Maintaining-Legacy-Support

jgambarios committed:
#31339 Add optional counter updates to parseLineResults method

Introduced a boolean parameter to determine whether counters should be updated in the parseLineResults method, ensuring flexibility for error handling scenarios. Removed redundant @NotNull annotations to simplify code. These changes improve clarity and functionality of the import utility.

jgambarios committed:
#31339 Add tracking and conversion of last inode in import process

This update introduces logic to track the last inode during imports and makes it accessible as an optional value. Adjustments include storing, retrieving, and converting the last inode to maintain consistency across processes. This enhances import result tracking and improves data handling.

jgambarios committed:
#31339 Update validation logic and assertion checks in ImportUtilTest

Adjusted validation parameters to align with updated requirements and fixed assertion to correctly reflect the expected error count. This ensures tests accurately validate warnings and errors, improving code reliability.

jgambarios committed:
#31339 Updating IT

jdotcms commented on dotCMS/src/main/java/com/dotcms/jobs/business/processor/impl/ImportContentletsProcessor.java:
```diff
@@ -325,11 +326,12 @@ public Map<String, Object> getResultMetadata(Job job) {
      * @param user             The user performing the import
      * @param progressCallback Callback for tracking import progress
      */
     private void handleImport(final boolean preview, final Job job, final File fileToImport,
```
more than three parameters smell bad, would consider to group into a bean

jgambarios replied:
This is related with how the `ImportUtil` already works and it is outside the scope of the ticket.

jdotcms commented on dotCMS/src/main/java/com/dotcms/jobs/business/processor/impl/ImportContentletsProcessor.java:
```diff
@@ -397,10 +402,12 @@ private Map<String, List<String>> processImport(final boolean preview, final Job
         Logger.info(this, String.format("-> Content Type: %s", contentType.variable()));
 
-        return ImportUtil.importFile(importId, currentSiteId, contentType.id(), fields, preview,
+        return ImportUtil.importFileResult(importId, currentSiteId, contentType.id(), fields,
```
Saml thing here, would use a bean with a builder pattern

jgambarios replied:
This is related with how the `ImportUtil` already works and it is outside the scope of the ticket.

jdotcms commented on dotCMS/src/main/java/com/dotcms/jobs/business/util/JobUtil.java:
```diff
@@ -160,4 +183,39 @@ public static <T extends JobEvent> void sendEvents(final T event) {
+            return Map.of();
+        }
+
+        JSONObject resultJsonObject = new JSONObject(asString);
```
this is a bad pattern you may be able to convert the object to map from jersey, this is spending so much cpu time and mem

jgambarios replied:
Fixed, now we are using `objectMapper.get().convertValue(toTransform, Map.class)`.

jdotcms commented on dotCMS/src/main/java/com/dotmarketing/util/ImportUtil.java:
```diff
@@ -226,123 +233,92 @@ public static HashMap<String, List<String>> importFile(
-        return importFile(importId, currentSiteId, contentTypeInode, keyfields, preview,
-                isMultilingual, user, language, csvHeaders, csvreader, languageCodeHeaderColumn,
-                countryCodeHeaderColumn, wfActionId, request, null);
+        ImportResult result = importFileResult(importId, currentSiteId, contentTypeInode,
```
again many arguments for a method

jgambarios replied:
This is related with how the `ImportUtil` already works and it is outside the scope of the ticket.

jdotcms commented on dotCMS/src/main/java/com/dotmarketing/util/ImportUtil.java:
```diff
@@ -226,123 +233,92 @@ public static HashMap<String, List<String>> importFile(
-        results.put(Contentlet.WORKFLOW_ACTION_KEY, new ArrayList<>());
-
-        Structure contentType = CacheLocator.getContentTypeCache().getStructureByInode (contentTypeInode);
+        Structure contentType = CacheLocator.getContentTypeCache()
```
why using legacy beans?

jgambarios replied:
One of the requirements for the ticket was to leave the logic exactly as it currently is.

jdotcms commented on dotCMS/src/main/java/com/dotmarketing/util/ImportUtil.java:
```diff
@@ -226,123 +233,92 @@ public static HashMap<String, List<String>> importFile(
-        for(Field field : FieldsCache.getFieldsByStructureInode(contentType.getInode())){
-            if(field.isUnique()){
+        // Get unique fields
+        for (Field field : FieldsCache.getFieldsByStructureInode(contentType.getInode())) {
```
this could be reduce to one line with stream + filter + collect

jgambarios replied:
Done with the stream changes.

jdotcms commented on dotCMS/src/main/java/com/dotmarketing/util/ImportUtil.java:
```diff
@@ -353,210 +329,438 @@ public static HashMap<String, List<String>> importFile(Long importId, String cur
-                                        }
-                                    }
-                                }
+                            Logger.debug(ImportUtil.class,
```
The methods in the class are so big

jgambarios replied:
This is related with how the `ImportUtil` already works and it is outside the scope of the ticket.

jdotcms commented on dotCMS/src/main/java/com/dotmarketing/util/importer/ImportResultConverter.java:
```diff
@@ -51,238 +38,182 @@ private ImportResultConverter() {
+        try {
+            final var preview = result.type().equals(OperationType.PREVIEW);
+
+            final HashMap<String, List<String>> legacyResults = new HashMap<>();
```
it would be better to create a bean for this, this is very old fashion

jgambarios replied:
One of the requirements for the ticket was to leave the logic exactly as it currently is, that why for the "transformer" we are using the same format/logic the import had when it is called from the existing struts action.

jgambarios committed:
#31339 Refactor JSON conversion logic in JobUtil.

Replaced JSONObject-based map transformation with ObjectMapper for better consistency and reduced dependency. Simplified `transformToMap` method implementation and updated the documentation accordingly. Additionally, minor code formatting improvements were made in ImportContentletsProcessor.

jgambarios committed:
Merge remote-tracking branch 'origin/main' into issue-31339-Migrate-ImportContentletsProcessor-to-Use-New-Import-Result-Format-While-Maintaining-Legacy-Support

jgambarios committed:
#31339 Simplify unique field retrieval logic in ImportUtil

Replaced manual iteration and filtering with stream-based processing to retrieve unique fields. This improves readability and maintains the same functionality with cleaner code.

jgambarios committed:
Merge remote-tracking branch 'origin/main' into issue-31339-Migrate-ImportContentletsProcessor-to-Use-New-Import-Result-Format-While-Maintaining-Legacy-Support

jgambarios committed:
#31339 Refactor to reuse 'data' variable in ImportResultConverter

Replaced repetitive calls to 'result.data()' with the 'data' variable to improve code readability and maintainability. This change reduces redundancy and ensures consistent access to the processed data throughout the affected methods.

jgambarios committed:
Merge remote-tracking branch 'origin/main' into issue-31339-Migrate-ImportContentletsProcessor-to-Use-New-Import-Result-Format-While-Maintaining-Legacy-Support

jgambarios commented on dotCMS/src/main/java/com/dotmarketing/util/importer/ImportResultConverter.java:
```diff
@@ -51,238 +38,182 @@ private ImportResultConverter() {
+                List<String> counters = legacyResults.get(KEY_COUNTERS);
+                counters.add("linesread=" + result.data().processed().parsedRows());
+                counters.add("errors=" + result.data().processed().failedRows());
+                counters.add("newContent=" + data.summary().toCreateContent());
```
Done
